src/repr/src/row.rs

```diff
@@ -427,7 +456,44 @@ fn read_untagged_bytes<'a>(data: &'a [u8], offset: &mut usize) -> &'a [u8] {
 /// This function is safe if the datum's length and contents were previously written by `push_lengthed_bytes`,
 /// and it was only written with a `String` tag if it was indeed UTF-8.
 unsafe fn read_lengthed_datum<'a>(data: &'a [u8], offset: &mut usize, tag: Tag) -> Datum<'a> {
-    let len = match tag {
+    let len = read_lengthened_datum_len(data, offset, tag);
+    let bytes = &data[*offset..(*offset + len)];
+    *offset += len;
+    match tag {
+        Tag::BytesTiny | Tag::BytesShort | Tag::BytesLong | Tag::BytesHuge => Datum::Bytes(bytes),
+        Tag::StringTiny | Tag::StringShort | Tag::StringLong | Tag::StringHuge => {
+            Datum::String(str::from_utf8_unchecked(bytes))
+        }
+        _ => unreachable!(),
+    }
+}
+
+/// Read a datum kind whose length is encoded in the row before its contents.
+///
+/// Updates `offset` to point to the first byte after the end of the read region.
+///
+/// # Safety
+///
+/// This function is safe if the datum's length and contents were previously written by `push_lengthed_bytes`,
+/// and it was only written with a `String` tag if it was indeed UTF-8.
+unsafe fn read_lengthed_datum_kind(data: &[u8], offset: &mut usize, tag: Tag) -> DatumKind {
```

Comment:
Why does this function (and `read_lengthened_datum_len` and `read_datum_kind`) need to be `unsafe`? I don't see them calling unsafe code. The "Safety" comments also don't really seem to apply. If you invoke this function on invalid data, it will produce garbage or panic, but it won't violate memory safety.